src/main/scala/wdlTools/types/Context.scala

```diff
@@ -59,55 +56,57 @@ case class Context(version: WdlVersion,
     this.copy(outputs = bindings)
   }
 
-  def bindVar(varName: String, wdlType: WdlType, textSource: TextSource): Context = {
+  def bindVar(varName: String, wdlType: WdlType): Either[Context, String] = {
     declarations.get(varName) match {
       case None =>
-        this.copy(declarations = declarations + (varName -> wdlType))
+        Left(this.copy(declarations = declarations + (varName -> wdlType)))
       case Some(_) =>
-        throw new TypeException(s"variable ${varName} shadows an existing variable",
-                                textSource,
-                                docSourceUrl)
+        Right(s"variable ${varName} shadows an existing variable")
     }
   }
 
-  def bindStruct(s: T_Struct, textSource: TextSource): Context = {
+  // add a bunch of bindings
+  // the caller is responsible for ensuring that each binding is either
+  // not a duplicate of an existing variables or that its type extends Invalid
+  def bindVarList(bindings: Map[String, WdlType]): Context = {
+    val both = declarations.keys.toSet intersect bindings.keys.toSet
+    val keep = bindings.filter {
+      case (name, _: Invalid) if both.contains(name) =>
+        // ignore duplicate var if it is Invalid
+        false
+      case (name, _) if both.contains(name) =>
+        throw new RuntimeException(
+            "Trying to bind variables that have already been declared with non-error types"
+        )
+      case _ => true
+    }
+    this.copy(declarations = declarations ++ keep)
+  }
+
+  def bindStruct(s: T_Struct): Either[Context, String] = {
```

Comment:
The `Either` construction is too heavy handed here. It is better to throw an exception than to change the type.